### PR TITLE
[Modelzoo]Rebuild MMoE to update API and Enable DeepRec Features

### PR DESCRIPTION
**Pull request by jmaksymc**
**WIP**

- Enable following DeepRec feature:
  - Enabled By Default:
      8) ~~Auto Micro Batch same with https://github.com/alibaba/DeepRec/issues/127~~
      9) ~~FusedEmbedding API, embedding fusion~~
      10) ~~Smart Stage same with https://github.com/alibaba/DeepRec/issues/122~~
      11) ~~Auto Graph Fusion https://github.com/alibaba/DeepRec/issues/144~~
      12) ~~CPU Memory Optimization:START_STATISTIC_STEP, STOP_STATISTIC_STEP, jemalloc~~
      14) ~~AdamAsync Optimizer~~
      15) ~~BF16~~
  - Disabled By Default:
         1) Embedding Variable **Not support EV yet**
         7) GRPC++ and StarServer **Implemented in the model but they are not supported yet**
         13) ~~Incremental Checkpoint~~
         14) ~~AdagradDecay~~
         2) EmbeddingVariable advanced features：Embedding Elimination **Not support EV yet**
         3) EmbeddingVariable advanced feature：Embedding Filter **Not support EV yet**
         4) Dynamic-dimension Embedding Variable - **There is no feature_column API**
         5) Adaptive Embedding - **ERROR:**  data.shape must start with partitions.shape, got data.shape = [18886], partitions.shape = [512]
          17) ~~WorkQueue~~
  - Other Features:
      6) Multi-Hash Variable - **There is no feature_column API**

### CODE REVIEW

File: modelzoo/MMoE/train.py [hunk 1]
```diff
@@ -0,0 +1,897 @@
...
+    
+    # used to add summary in tensorboard
+    def _add_layer_summary(self, value, tag):
+        tf.summary.scalar('%s/fraction_of_zero_values' % tag,
+                        tf.nn.zero_fraction(value))
+        tf.summary.histogram('%s/activation' % tag, value)
+    
+    def _make_scope(self, name, bf16, part):
+        if(bf16):
+            return tf.variable_scope(name, partitioner=part, reuse=tf.AUTO_REUSE).keep_weights()
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:162]: keep_weights(dtype=tf.float32), this makes it clearer what the function does.

File: modelzoo/MMoE/train.py [hunk 2]
```diff
@@ -0,0 +1,897 @@
...
+        files = filename
+
+    # Extract lines from input files using the Dataset API.
+    dataset = tf.data.TextLineDataset(files)
+    dataset = dataset.shuffle(buffer_size=400000,
+                              seed=args.seed)  # set seed for reproducing
+    dataset = dataset.repeat(num_epochs)
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(parse_csv, num_parallel_calls=28)
+    dataset = dataset.prefetch(32)
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:345]: prefetch(2), keep aligned with other models in modelzoo.

File: modelzoo/MMoE/train.py [hunk 3]
```diff
@@ -0,0 +1,897 @@
...
+    ) if args.micro_batch and not args.tf else args.batch_size
+
+    if args.steps == 0:
+        no_of_epochs = 3
+        train_steps = math.ceil(
+            (float(no_of_epochs) * no_of_training_examples) / batch_size)
+    else:
+        no_of_epochs = math.ceil(
+            (float(batch_size) * args.steps) / no_of_training_examples)
+        train_steps = args.steps
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:600]: Please be sure that by python train.py, the model will be trained for at least 3000 steps. Because the test script in DeepRec will run python train.py and collect the gsteps between 2000~3000. The script is under DeepRec/tests/model_benchmark.

File: modelzoo/MMoE/train.py (hunk 1 above)
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Changed

File: modelzoo/MMoE/train.py (hunk 3 above)
Comment (jmaksymc) [reply modelzoo/MMoE/train.py:600]: Changed

File: modelzoo/MMoE/train.py (hunk 2 above)
Comment (jmaksymc) [reply modelzoo/MMoE/train.py]: Changed

File: modelzoo/MMoE/train.py (hunk 3 above)
Comment (Duyi-Wang) [reply modelzoo/MMoE/train.py:600]: Please change deafult steps by `no_of_epochs`

**Comment by Duyi-Wang** [comment]
Please update README、distribute training files and add DBMTL in modelzoo/README. Adaptive Embedding issue has been submitted to Ali https://github.com/alibaba/DeepRec/issues/236

**Comment by shanzhou2186** [comment]
Please help to change the steps number and update the performance data  of Alicloud CPX in JIRA and ignore the accuracy for such a small dataset. Our SPR platform ran into issues, which hasn't been fixed yet.  Then we can push the PR to Alibaba.

**Comment by Duyi-Wang** [comment]
- Please add a directory structure at the start of REAMDE.
- Please update distribute training files, keep align with other models.

File: modelzoo/MMoE/train.py [hunk 4]
```diff
@@ -0,0 +1,897 @@
...
+  return l2
+
+class MMOE():
+    def __init__(self,
+                 input,
+                 feature_column,
+                 num_experts,
+                 expert_hidden_units,
+                 towers,
+                #  l2_scale,
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:119]: clean

File: modelzoo/MMoE/train.py [hunk 5]
```diff
@@ -0,0 +1,897 @@
...
+            raise ValueError("Dataset is not defined.")
+        self._feature = input[0]
+        self._label = input[1]
+
+        self._feature_column = feature_column
+        self._num_experts = num_experts
+        self._expert_hidden_units = expert_hidden_units
+        self._towers = towers
+
+        # self.__l2_regularization = l2_regularizer(l2_scale) if l2_scale else None
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:137]: clean

File: modelzoo/MMoE/train.py [hunk 6]
```diff
@@ -0,0 +1,897 @@
...
+            for i in range(1, self._num_experts + 1):
+                with tf.variable_scope(f'expert_{i}'):
+                    expert_features = input_emb
+
+                    for layer_id, num_hidden_units in enumerate(self._expert_hidden_units):
+                        with tf.variable_scope(f'expert_{i}_layer_{layer_id}', reuse=tf.AUTO_REUSE) as expert_layer_scope:
+                            expert_features = tf.layers.dense(expert_features,
+                                                            units=num_hidden_units,
+                                                            activation=None,
+                                                            # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:206]: clean

File: modelzoo/MMoE/train.py [hunk 7]
```diff
@@ -0,0 +1,897 @@
...
+
+            towers=[]
+            for tower in self._towers:
+                tower_name = tower[0]
+                hidden_units = tower[2]
+                with tf.variable_scope(f'{tower_name}_gate', reuse=tf.AUTO_REUSE) as gate_scope:
+
+                    gate = tf.layers.dense(input_emb,
+                                        units=self._num_experts,
+                                        # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:222]: clean

File: modelzoo/MMoE/train.py [hunk 8]
```diff
@@ -0,0 +1,897 @@
...
+                                                            units=num_hidden_units,
+                                                            # kernel_regularizer=self.__l2_regularization,
+                                                            name=f'{tower_layer_scope.name}/dense')
+                            tower_features = DNN_ACTIVATION(tower_features,
+                                                            name=f'{tower_layer_scope.name}/act')
+                            self._add_layer_summary(tower_features, tower_layer_scope.name)
+                    final_tower_predict = tf.layers.dense(inputs=tower_features,
+                                                        units=1,
+                                                        activation=None,
+                                                        # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:246]: clean

File: modelzoo/MMoE/train.py [hunk 9]
```diff
@@ -0,0 +1,897 @@
...
+
+                    tower_input = tf.multiply(experts_features, gate)
+                    tower_input = tf.reduce_sum(tower_input, axis=1)
+
+                    tower_features = tower_input
+                    for layer_id, num_hidden_units in enumerate(hidden_units):
+                        with tf.variable_scope(f'{tower_name}_layer_{layer_id}', reuse=tf.AUTO_REUSE) as tower_layer_scope:
+                            tower_features = tf.layers.dense(tower_features,
+                                                            units=num_hidden_units,
+                                                            # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:238]: clean

File: modelzoo/MMoE/train.py [hunk 10]
```diff
@@ -0,0 +1,897 @@
...
+                            tower_features = DNN_ACTIVATION(tower_features,
+                                                            name=f'{tower_layer_scope.name}/act')
+                            self._add_layer_summary(tower_features, tower_layer_scope.name)
+                    final_tower_predict = tf.layers.dense(inputs=tower_features,
+                                                        units=1,
+                                                        activation=None,
+                                                        # kernel_regularizer=self.__l2_regularization,
+                                                        name=f'{tower_name}_output')
+                    self._add_layer_summary(final_tower_predict, f'{tower_name}_output')
+                    # final_tower_predict = tf.math.sigmoid(final_tower_predict, f'{tower_name}_output')
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:249]: clean

File: modelzoo/MMoE/train.py [hunk 11]
```diff
@@ -0,0 +1,897 @@
...
+            #         '''
+            #         column = tf.feature_column.categorical_column_with_adaptive_embedding(
+            #             column_name,
+            #             hash_bucket_size=HASH_BUCKET_SIZES[column_name],
+            #             dtype=tf.string,
+            #             ev_option=ev_opt)
+            #     elif args.dynamic_ev:
+            #         '''Dynamic-dimension Embedding Variable'''
+            #         print("Dynamin-dimension Embedding Variable isn't really enabled in model.")
+            #         sys.exit()
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:472]: clean

File: modelzoo/MMoE/train.py [hunk 12]
```diff
@@ -0,0 +1,897 @@
...
+                    [model.acc_op, model.auc_op, merged])
+                writer.add_summary(events, _in)
+                print("Evaluation complete:[{}/{}]".format(_in, steps))
+                print("ACC = {}\nAUC = {}".format(eval_acc, eval_auc))
+
+def main(tf_config=None, server=None):
+        # check dataset and count data set size
+    print("Checking dataset...")
+    train_file = os.path.join(args.data_location, 'train')
+    test_file = os.path.join(args.data_location, 'eval')
```
Comment (Duyi-Wang) [inline modelzoo/MMoE/train.py:578]: train_file = os.path.join(args.data_location, 'taobao_train_data')
    test_file = os.path.join(args.data_location, 'taobao_test_data')
